docs/contribute/error_handling.rst

```diff
@@ -0,0 +1,118 @@
+
+.. code:: c
+
+  // src/api_test.cc
+  void ErrorTest(int x, int y) {
+    CHECK_EQ(x, y) << "ValueError: expect x and y to be equal."
+    if (x == 1) {
+      LOG(FATAL) << "InternalError: cannot reach here";
+    }
+  }
+
+The above function is registered as PackedFunc into the python frontend,
+under the name ``tvm._api_internal._ErrorTest``.
+Here is what will happen if we call the registered function:
+
+.. code::
+
+  >>> import tvm
+  >>> tvm._api_internal._ErrorTest(0, 1)
+  Traceback (most recent call last):
+    File "<stdin>", line 1, in <module>
+    File "/path/to/tvm/python/tvm/_ffi/_ctypes/function.py", line 190, in __call__
+      raise get_last_ffi_error()
+  ValueError: Traceback (most recent call last):
+    [bt] (3) /path/to/tvm/build/libtvm.so(TVMFuncCall+0x48) [0x7fab500b8ca8]
+    [bt] (2) /path/to/tvm/build/libtvm.so(+0x1c4126) [0x7fab4f7f5126]
+    [bt] (1) /path/to/tvm/build/libtvm.so(+0x1ba2f8) [0x7fab4f7eb2f8]
+    [bt] (0) /path/to/tvm/build/libtvm.so(+0x177d12) [0x7fab4f7a8d12]
+    File "/path/to/tvm/src/api/api_test.cc", line 80
+  ValueError: Check failed: x == y (0 vs. 1) : expect x and y to be equal.
+  >>>
+  >>> tvm._api_internal._ErrorTest(1, 1)
+  Traceback (most recent call last):
+    File "<stdin>", line 1, in <module>
+    File "/path/to/tvm/python/tvm/_ffi/_ctypes/function.py", line 190, in __call__
+      raise get_last_ffi_error()
+  tvm.error.InternalError: Traceback (most recent call last):
+    [bt] (3) /path/to/tvm/build/libtvm.so(TVMFuncCall+0x48) [0x7fab500b8ca8]
+    [bt] (2) /path/to/tvm/build/libtvm.so(+0x1c4126) [0x7fab4f7f5126]
+    [bt] (1) /path/to/tvm/build/libtvm.so(+0x1ba35c) [0x7fab4f7eb35c]
+    [bt] (0) /path/to/tvm/build/libtvm.so(+0x177d12) [0x7fab4f7a8d12]
+    File "/path/to/tvm/src/api/api_test.cc", line 83
+  InternalError: cannot reach here
+  TVM hint: You hit an internal error. Please open a thread on https://discuss.tvm.ai/ to report it.
+
+As you can see in the above example, TVM's ffi system combines
+both the python and c++'s stacktrace into a single message, and generate the
+corresponding error class automatically.
+
+
+How to choose an Error Type
+---------------------------
+You can go through the error types are listed below, try to us common
+sense and also refer to the choices in the existing code.
+We try to keep a reasonable amount of error types.
+If you feel there is a need to add a new error type, do the following steps:
+
+- Send a RFC proposal with a description and usage examples in the current codebase.
+- Add the new error type to :any:`tvm.error` with clear documents.
+- Update the list in this file to include the new error type.
+- Change the code to use the new error type.
+
+We also recommend to use less abstraction when creating the short error messages.
+The code is more readable in this way, and also opens path to craft specific
+error messages when necessary.
+
+.. code:: python
+
+   def preferred():
```

Comment:
I'm not completely sure about which of these are preferred. Is only the first option valid? Option 2 is not preferred, right?

Reply:
more discussions in https://github.com/dmlc/tvm/issues/2279, i think this is something that maybe worth some more discussion